update: encrypt.md (#1716)
juicedata/juicefs: docs/en/security/encrypt.md

# Data Encryption

JuiceFS provides data encrytion from two aspects:

1. Data Encryption In Transit
2. Data Encryption At Rest

## Data Encryption In Transit

Running JuiceFS generally involves the network connection between database and object storage, which is determined by the architecture of JuiceFS. As long as the servers support encryption connections, JuiceFS can be accessed through the encrypted channel.

### Connect to objetc storage via HTTPS

Public cloud object storage generally supports both HTTP and HTTPS. If no scheme is specified, JuiceFS uses HTTPS by default. For example, the client will identify the bucket in following command as `https://myjfs.s3.ap-southeast-1.amazonaws.com`.

```shell
juicefs format --storage s3
--bucket myjfs.s3.ap-southeast-1.amazonaws.com
...
```

In the case where server and object storage run on the same VPC network, explicitly set the URL scheme to `http` if you don't need an encrypted connection, e.g., `--bucket http://myjfs.s3.ap-southeast-1.amazonaws.com`.

### Connect to database via TLS/SSL

> For [all the supported metadata engines](../reference/how_to_setup_metadata_engine.md), as long as the database supports encryption and has been configured with encryption such as TLS/SSL, JuiceFS can connect to the database through its encrytped channel. For instance, a Redis database configured with TLS can use `redis://` for connecting.

Reviewer: There should be have two `s` here.
```suggestion
For [all the supported metadata engines](../reference/how_to_setup_metadata_engine.md), as long as the database supports encryption and has been configured with encryption such as TLS/SSL, JuiceFS can connect to the database through its encrytped channel. For instance, a Redis database configured with TLS can use `rediss://` for connecting.
```

```shell
juicefs format --storage s3
--bucket myjfs.s3.ap-southeast-1.amazonaws.com
"rediss://myredis.ap-southeast-1.amazonaws.com:6379/1" myjfs
```

----

JuiceFS encrypts data during transmission over the network to prevent unauthorized users from eavesdropping on network traffic.

JuiceFS clients always use HTTPS to upload data to the Object Storage Service, except for the following cases.

- Uploading to Alibaba Cloud OSS using internal endpoints
- Uploading to UCloud US3 using internal endpoints

-----

## Data Encryption At Rest

JuiceFS supports Data Encryption At Rest. All data will be encrypted first before being uploaded to object storage. This allows JuiceFS to effectively prevent data leakage.

> Data Encryption At Rest of JuiceFS adopts industry-standard encryption methods (AES-GCM and RSA). User only needs to provide a private key or password while creating the file system, and the private key can be provided by setting the environment variable `JFS_RSA_PASSPHRASE`. For use, mount point is totally transparent to the client, i.e., access to file system will not be affected by encryption and decryption processes.

Reviewer: `or` -> `with`

```suggestion
Data Encryption At Rest of JuiceFS adopts industry-standard encryption methods (AES-GCM and RSA). User only needs to provide a private key with a password while creating the file system, and the private key can be provided by setting the environment variable `JFS_RSA_PASSPHRASE`. For use, mount point is totally transparent to the client, i.e., access to file system will not be affected by encryption and decryption processes.
```

![Encryption At-rest](../images/encryption.png)

The detailed process of data encryption is as follows:

- Before writing to an object storage, data blocks are compressed using LZ4 or ZStandard.
- A random 256-bit symmetric key `S` and a random seed `N` are generated for each data block.
- Each data block is encrypted using `S` and `N` based on AES-GCM encryption yielding `encrypted_data`.
- To avoid the symmetric key `S` from being transmitted in clear text over the network, the symmetric key `S` is encrypted with the RSA key `M` to obtain the ciphertext `K`.
- The encrypted data `encrypted_data`, the ciphertext `K`, and the random seed `N` are combined into an object and then written to the object storage.

The steps for decrypting the data are as follows:

- Read the entire encrypted object (it may be a bit larger than 4MB).
- Parse the object data to get the ciphertext `K`, the random seed `N`, and the encrypted data `encrypted_data`.
- Decrypt `K` with RSA key to get symmetric key `S`.
- Decrypt the data `encrypted_data` based on AES-GCM using `S` and `N` to get the data block plaintext.
- Decompress the data block.

### Enable Data Encryption At Rest

:::note

> Data Encryption At Rest must be enabled when creating file system. The file system that has areadly enabled Data Encryption At Rest does not need to enable it again. 

Author: ```suggestion
Data Encryption At Rest must be enabled when creating file system. The file system that was created without Data Encryption At Rest enabled cannot enable it later.
```

:::

There are following steps to enable Data Encryption At Rest:

1. Create a RSA private key
2. Run the encrypted file system that is created by the RSA private key
3. Mount the file system

#### Step 1: create a RSA private key

The RSA private key plays a crtical role in Data Encryption At Rest, usually manually generated by OpenSSL. The following command generates a 2048-bit RSA private key with a file name of `my-priv-key.pem` in the currect directory using aes256 algorithm.

```shell
openssl genrsa -out my-priv-key.pem -aes256 2048
```

> Because of the aes256 algorithm, a passphrase (at least 4 bits) is required for this RSA private key. The passphrase can be considered as a password for encrypting the file of the RSA private key, which is also the last layer of security assurance for the RSA private key.

Reviewer 2: ```suggestion
Since the aes256 algorithm is in use, you will be prompted to enter a passphrase (at least 4 bits) to protect your private key. The passphrase can be considered as a password for encrypting the file of the RSA private key, which is also the last layer of security assurance for the RSA private key.
```

:::caution

> The security of RSA private key is critical when Data Encryption At Rest encryption is enabled. Once the key is compromised, the data can confront security risk. If the key is lost, **all the encrypted data will be lost and cannot be recovered**!

Reviewer 2: ```suggestion
The security of RSA private key is crucial when Data Encryption At Rest encryption is enabled. The leak of the key will put the data into a serious security risk. If the key is lost, **all the encrypted data will be lost and cannot be recovered**!
```

:::

#### Step 2: create an encrypted file system

> The option `--encrypt-rsa-key` is required to specify RSA private key when creating an encrypted file system. The provided private key content will be written to the metadata engine. Since passphrase is mandatory in the ase256-encrypted RSA private key, the environmental variable `JFS_RSA_PASSPHRASE` is required to specify the passphrase of the private key before creaging and mounting file system.

Reviewer 2: ```suggestion
The option `--encrypt-rsa-key` is required to specify RSA private key when creating an encrypted file system. The provided private key content will be written to the metadata engine. Since passphrase is mandatory in the ase256-encrypted RSA private key, the environment variable `JFS_RSA_PASSPHRASE` is required to specify the passphrase of the private key before creating and mounting file system.
```

> 1. Set passphrase using environmental variable

Reviewer 2: ```suggestion
1. Set passphrase using environment variable
```

    ```shell
    export JFS_RSA_PASSPHRASE=the-passwd-for-rsa
    ```

2. Create file system

    ```shell
    juicefs format --storage s3 \
    --encrypt-rsa-key my-priv-key.pem \
    ...
    ```

#### Step 3: mount file system

> There is no need to specify options while mounting an encrypted file system. However, the passphrase of the private key needs to be set before mounting using environmental variable.

Reviewer 2: ```suggestion
There is no need to specify extra options while mounting an encrypted file system. However, the passphrase of the private key needs to be set before mounting using environment variable.
```

> 1. Set passphrase using environmental variable

Reviewer 2: ```suggestion
1. Set passphrase using environment variable
```

### Performance

> TLS, HTTPS, and AES-256 are implemented very efficiently in modern CPUs. Therefore, enabling encryption does not have a significant impact on file system performance. Note that RSA algorithms are relatively slow, especially in the decryption process. Additionally, it is recommended to use 2048-bit RSA keys for storage encryption, and using 4096-bit keys may have a significant impact on reading performance.

Reviewer 2: ```suggestion
TLS, HTTPS, and AES-256 are implemented very efficiently in modern CPUs. Therefore, enabling encryption does not have a significant impact on file system performance. Because of the relatively low performance of RSA algorithm, it is recommended to use 2048-bit RSA keys for storage encryption, and using 4096-bit keys may have a significant impact on reading performance.
```
